Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -151,6 +151,47 @@ TEST_F(SerializerTest, EncodesIntegersModelToBytes) {
   }
 }
 
+TEST_F(SerializerTest, EncodesStringModelToBytes) {
+  struct TestCase {
+    std::string value;
+    std::vector<uint8_t> bytes;
+  };
+
+  std::vector<TestCase> cases{
+      // TEXT_FORMAT_PROTO: 'string_value: ""'
+      {"", {0x8a, 0x01, 0x00}},
+      // TEXT_FORMAT_PROTO: 'string_value: "a"'
+      {"a", {0x8a, 0x01, 0x01, 0x61}},
+      // TEXT_FORMAT_PROTO: 'string_value: "abc def"'
+      {"abc def", {0x8a, 0x01, 0x07, 0x61, 0x62, 0x63, 0x20, 0x64, 0x65, 0x66}},
+      // TEXT_FORMAT_PROTO: 'string_value: "æ"'
+      {"æ", {0x8a, 0x01, 0x02, 0xc3, 0xa6}},
+      // TEXT_FORMAT_PROTO: 'string_value: "\0\ud7ff\ue000\uffff"'
+      // Some magic: Due to the unicode escapes (\u) and
```

Comment:
Nit: feel free to disagree, but I don't think this is magic. It's valid to embed a universal character name inside a `char` string literal, it will map to several characters if necessary, so I'm not sure if there's any narrowing involved. Consider rephrasing to something like:

"Each one of the three embedded universal character names (\u-escaped) maps to three `char`s, so the total length of the string literal is 10 (ignoring the terminating null), and the resulting string literal is the same as '\0\xed\x9f\xbf\xee\x80\x80\xef\xbf\xbf'"

or similar.

Reply:
I like your phrasing better. (Done) The thing that confused me initially was the \0 prefix. I didn't look close enough to realize that's a null, rather than just another unicode escape character, and was surprised when my encoding function encoded 0 bytes. :)

Reply:
Ah, yeah, `std::string` can contain embedded nulls, which can be a gotcha when using `c_str`.

Reply:
In my case, it was actually the other way around. i.e. std::string("\0ABCDEFG") contains an empty string and doesn't have any embedded nulls in it. (The ctor stops processing the input char* as soon as it sees the null at position 0... unless you also indicate the size.)

Reply:
Right, that constructor presumes a null-terminated string, because there's no way to figure out the role of a given null. I had a funny quirk reimplementing some Objective C parsing in C++: when parsing an std::string, I had to make sure parsing stops upon encountering '\0' even if it comes before the last character in the string is reached, for consistency with existing behavior (this might be a little paranoid, though).